test/libsolidity/SMTCheckerTest.cpp

```diff
@@ -32,8 +32,13 @@ SMTCheckerTest::SMTCheckerTest(std::string const& _filename):
 	universalCallback(nullptr, smtCommand)
 {
 	auto contract = m_reader.stringSetting("SMTContract", "");
-	if (!contract.empty())
+	auto maybeContracts = ModelCheckerContracts::fromString(contract);
+	if (maybeContracts)
+		m_modelCheckerSettings.contracts = *maybeContracts;
+	else if (!contract.empty())
 		m_modelCheckerSettings.contracts.contracts[""] = {contract};
+	else
+		BOOST_THROW_EXCEPTION(std::runtime_error("Invalid contract specified."));
```

Comment:
I'm not sure why `fromString()` is even defined as returning `optional`, because it never returns `nullopt`. You can just assert that, or better, remove the optional and the error.

Reply:
BTW if you do end up keeping the error for some reason, I'd at least mention the setting name. These exceptions do not point at a specific place in the test that's wrong so it may be confusing otherwise.

Reply:
`fromString` has a path when it returns an empty optional. I have made that explicit with `std::nullopt`. Previously, `{}` was used.

Reply:
Ah, right. I missed that. This logic still seems a bit off to me though. Especially the fact that an exception sounds like it should be thrown on invalid names but actually only happens when the entire setting is left empty.

For example, I think that something like this will fail the first check and enter the second branch:
```
// SMTContract: a.sol:A, B, c.sol:C
```
which means that we'll try to use `a.sol:A, B, c.sol:C` as a contract name, without reporting an error. A more robust way to do it would be to check whether the string contains `:`, `,` or whitespace and report the error when it does, not just when it's empty.

Reply:
BTW, note that an empty source name is actually something that can happen. You can't do this on the CLI, but in Standard JSON source names can be arbitrary strings, including empty ones. There are some quirks around their behavior (#15747) and I think we'd be better off disallowing them, but for the time being they're technically valid.

And they can contain `:` and `,` even on the CLI (as long as the file system allows this). Only contract names are guaranteed to be identifiers.

Reply:
I'll have to double-check this, but you are probably right.
I'll see if I can make the parsing robust easily.

> This does not matter all that much for testing, but AFAIK fromString() is also used for parsing the value of the option in normal operation, so it means that SMTChecker probably can't process contracts in some weirdly named files.

I would prefer to ignore weirdly named files for now, unless someone complains.

Reply:
> I would prefer to ignore weirdly named files for now, unless someone complains.

I really think we should be consistent about it. If the rest of the compiler works like this, the SMTChecker should too.

Though of course it's a very low priority thing. From my point of view it's a bug, but it's perfectly fine to just file it and only get to it when there's nothing more important to work on.

I'd also like to make the path situation saner in one of the breaking release so not jumping right at it is actually good, but I can't say when that will happen, and some of these things will still be problematic (e.g. I don't see us really disallowing commas or colons in paths).

Reply:
I have created https://github.com/ethereum/solidity/issues/15945 to keep track of this issue.

I think this was the last issue regarding this PR?

Reply:
Yes. Still does not seem resolved though. I expected you'd replace the check for empty with one for strings containing `:`, `,` or whitespace.

I mean, I'm not going to block the PR because of that, but it seems straightforward to address. To be clear I'm not talking about a fix for #15945, just adjusting this one place that parses the test setting.

Reply:
Sorry, I missed that objection.
I guess I don't quite understand what you would like to see there.
This is about our test setup.
Previously, any value for `SMTContract` was accepted, unless it was an empty string, in which case it was silently ignored.
I am adding an option to refer to a specific contract in a specific file in a multi-file scenario and I also wanted to turn the empty string case into error.
It would make sense to check if the string is a valid contract name, but that seems more general than what you suggested, no?

Reply:
After thinking about it more, I added a stronger check for a valid contract name.
Is that what you had in mind, @cameel?